.github/workflows/ubuntu-20.04.yml

```diff
@@ -107,22 +81,28 @@ jobs:
       shell: bash
       run: echo ::set-output name=main-sha::$(git rev-parse refs/heads/main)
     - id: UbuntuVersionMatrix
-      name: Setup UbuntuVersion matrix
+      name: Setup UbuntuVersionMatrix
       shell: bash
       # For PRs, we only check a subset of the UbuntuVersions.
       # For everything else (nightly, Merge, etc.) we run all of them.
-      run: |
-        if [ "$GITHUB_REF" == 'refs/heads/main' ] || [ "$GITHUB_EVENT_NAME" == "schedule" ]; then
-          echo '::set-output name=UbuntuVersionMatrix::[16.04, 18.04, 20.04]'
+      # Adapted to support templating of this file so we can run each version in a separate workflow
+      # for easier retries rather than with matrix rules that require retrying the entire set
+      # (though we retain the matrix rules for convenience for now).
+      run: |
+        if [ "$GITHUB_REF" == 'refs/heads/main' ] || [ "$GITHUB_EVENT_NAME" == 'schedule' ]; then
+          echo '::set-output name=UbuntuVersionMatrix::["20.04"]'
+        elif ! echo '20.04' | egrep -q '^18.04$'; then
+          echo '::set-output name=UbuntuVersionMatrix::["20.04"]'
         else
-          echo '::set-output name=UbuntuVersionMatrix::[16.04, 20.04]'
+          echo '::set-output name=UbuntuVersionMatrix::["SKIP_ME"]'
         fi
```

Comment:
Previously we used dynamically generated lists of versions (based on whether we were in PR mode or CI mode) in a matrix for the later rules to use.

Now we template those out one for each version, but that means we need a non-empty value to check for when we have nothing to do during PR mode (e.g. skip Ubuntu 18.04 checks).
This also means lots of `if: ` conditional job checks on the remainder of the job templates.